[Li Guohuang] ip

guohuang88 opened the pull request:
# GhBot
GhBot is an efficient chat bot that remembers your tasks for you!
GhBot currently supports the following commands:

1. **Listing the tasks**
    - `list`

2. **Adding new tasks**
There are three different ways you can add a task.
    - `todo <task name>`
    - `deadline <task name> /by <date&time>`
    - `event <task name> /from <date&time> /to <date&time>`

3. **Deleting a task**
    - `delete <index of the task>`
    
4. **Marking a completed task** 
    - `mark <index of the task>`

5. **Unmarking an uncompleted task**
    - `unmark <index of the task>`

6. **Finding tasks with keyword**
    - `find <keyword>`

> [!IMPORTANT]
> Upcoming features include:
> - [x] ~~Finding tasks with keyword (completed)~~
> - [x] GUI using JavaFX technology (working in progress)

You can try out this chat bot [here](https://github.com/guohuang88/ip) :sunglasses:
If not you can:
```
git clone https://github.com/guohuang88/ip.git
```
If you are a Java programmer, you can practice Java by running the `main` method below!
```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

guohuang88 committed:
Completed Level-0

guohuang88 committed:
Completed Level-1

guohuang88 committed:
Completed Level-2

guohuang88 committed:
Completed Level-3

guohuang88 committed:
Completed Level-4

guohuang88 committed:
Completed A-TextUiTesting

guohuang88 committed:
Completed Level-5

guohuang88 committed:
Completed Level-6

guohuang88 committed:
Completed A-Enums

guohuang88 committed:
Add save task ability, load data ability and update javadoc

guohuang88 committed:
Add JavaDoc for DukeException class

guohuang88 committed:
Merge branch 'branch-Level-7'

guohuang88 committed:
Modify Duke class input to understand date/time instead of string

guohuang88 committed:
Minor comment fix

guohuang88 committed:
Merge tag 'branch-Level-8'

guohuang88 committed:
Modify handling of file in Duke to Storage

guohuang88 committed:
Modify handling of input in Duke class to Ui class

guohuang88 committed:
Relocate the parse method from Duke to Parser

guohuang88 committed:
Add TaskList class

guohuang88 committed:
Minor refactoring of code

guohuang88 committed:
Merge tag 'A-MoreOOP'

guohuang88 committed:
Add all files into ghbot package

guohuang88 committed:
Minor text file changes

guohuang88 committed:
Merge branch 'branch-A-Packages'

guohuang88 committed:
Merge commit 'a6f73244a20539e20c2189973ebf597df34ef631'

guohuang88 committed:
Add gradle file

guohuang88 committed:
Add JUnit test files

guohuang88 committed:
Minor JavaDoc fix

guohuang88 committed:
Merge branch 'branch-A-JUnit'

guohuang88 committed:
Update build.gradle to create jar file

guohuang88 committed:
Add JavaDoc to the method in the classes

guohuang88 committed:
Update some changes regarding CodingStandard

guohuang88 committed:
Add find feature for chatbot

guohuang88 committed:
Add header for test file

guohuang88 committed:
Merge branch 'branch-A-JavaDoc'

guohuang88 committed:
Merge branch 'branch-A-CodingStandard'

guohuang88 committed:
Merge branch 'branch-Level-9'

guohuang88 committed:
Modify OOP structure for the project

guohuang88 committed:
Add JavaDoc to the new classes

guohuang88 committed:
Update minor changes in JavaDoc

guohuang88 committed:
Merge branch 'branch-A-Improve-OOP'

guohuang88 committed:
Modify the naming of the constant names

guohuang88 committed:
Update the test file to cater to the most recent implementation

guohuang88 committed:
Update styling changes after running check-style

guohuang88 committed:
Modify minor JavaDoc changes in TaskList class

guohuang88 committed:
Merge branch 'branch-A-CheckStyle'

jmsandiegoo commented on src/main/java/ghbot/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+package ghbot;
```
Good job with packaging classes together 👍

jmsandiegoo commented on src/main/java/ghbot/Event.java:
```diff
@@ -0,0 +1,40 @@
+     * @return a string.
+     */
+    @Override
+    public String toFile() {
```
Excellent use of proper method naming with the use of 'verbs' throughout the codebase 😄

jmsandiegoo commented on src/main/java/ghbot/GhBot.java:
```diff
@@ -0,0 +1,38 @@
+    public static void main(String[] args) {
+        String fileName = "./data/history.txt";
+        Storage storage = new Storage(fileName);
+        TaskList taskLst = new TaskList(storage.getInputFromFile());
```
It seems good but would it be better if taskLst can be spelled fully?

jmsandiegoo reviewed (commented):
LGTM 💯 overall codebase is neat! Just a nit to maybe consider.

guohuang88 committed:
Add JavaFX GUI to the project

guohuang88 committed:
Update JavaDoc

guohuang88 committed:
Merge tag 'branch-Level-10'

guohuang88 committed:
Update project by packaging classes in an orderly manner

yapxuanxuan commented on src/main/java/ghbot/GhBot.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class GhBot {
+    private Storage storage;
+    private TaskList taskLst;
```
typo? I like that the variables are written in camelCase but perhaps could rename as taskList.

yapxuanxuan commented on src/main/java/ghbot/GhBot.java:
```diff
@@ -0,0 +1,56 @@
+     * @param filename The name of the file.
+     */
+    public GhBot(String filename) {
+        this.storage = new Storage(filename);
```
Perhaps could omit the 'this' keyword? I previously asked my TA about it and she mentioned this:
I got a question from one of your classmates regarding the this keyword. Use of this is discouraged, potentially to avoid unnecessary noise in the code. Try to avoid using this except when it is shadowed by a method or constructor parameter. This is an advanced rule that is optional to follow for CS2103T.

yapxuanxuan commented on src/main/java/ghbot/GhBot.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    /**
+     * You should have your own function to generate a response to user input.
```
Is this meant to be the java documentation for this method? JavaDoc does not seem to be properly written. Perhaps, could remove this block.

yapxuanxuan commented on src/main/java/ghbot/executor/FindExecutor.java:
```diff
@@ -0,0 +1,48 @@
+        for (int i = 0; i < this.taskList.taskSize(); i++) {
+            String[] words = this.taskList.getTask(i).toString().split(" ");
+            for (int j = 0; j < words.length; j++) {
+                if (this.description.equalsIgnoreCase(words[j])) {
```
I like that you used this function equalsIgnoreCase to ensure that the code works even with different case words. I did not know about this function so I learnt about it 👍

yapxuanxuan commented on src/main/java/ghbot/executor/FindExecutor.java:
```diff
@@ -0,0 +1,48 @@
+            }
+        }
+        if (matchedLst.size() > 0) {
+            this.executeStr = "Here are the list of tasks that matches the keyword:";
```
Perhaps english grammar mistake? Could be "Here are the list of tasks that match the keyword:" instead.

yapxuanxuan commented on src/main/java/ghbot/javafx/MainWindow.java:
```diff
@@ -0,0 +1,84 @@
+     * the dialog container. Clears the user input after processing.
+     */
+    @FXML
+    private void handleUserInput() {
```
i like that the method names are meaningful.

yapxuanxuan commented on src/main/java/ghbot/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class Task {
+    private String description;
+    private boolean isDone;
```
I like that the boolean variable is named to sound like a boolean

yapxuanxuan commented on src/main/java/ghbot/task/TaskList.java:
```diff
@@ -0,0 +1,60 @@
+ * Contains a list of tasks.
+ */
+public class TaskList {
+    private List<Task> tasks;
```
I like that the collection of tasks is named in plural form

guohuang88 committed:
Update build.gradle file to create the latest version of the jar file

guohuang88 committed:
Parser class: add assert statement in the else clause of parse() method

The parse() method does not ensure whether other inputs are available or
not available.

Add assert statement in the else clause.

Adding an assert statement checks that the else clause is unreachable
during testing.

Let's
* add a comment to tell where are the inputs being handled
* add an assert statement in the else clause to ensure that the else
clause is unreachable during testing

guohuang88 committed:
Merge pull request #1 from guohuang88/branch-A-Assertions

Parser class: add assert statement in the else clause of parse() method

guohuang88 committed:
Storage class: improve code quality

Storage class has a very long method and there are magic literals.

Long methods and magic literals can cause readability issues.

Extracting code for the different type of tasks such as Todo, Deadline
and Event into smaller methods and changing magic literals to constants
to improve readability.

Let's
* create small methods for the different type of tasks to enhance
readability of the long method
* create constants with useful naming convention to eliminate magic
literals

guohuang88 committed:
Merge pull request #2 from guohuang88/branch-A-CodeQuality

Storage class: improve code quality

guohuang88 committed:
FindExecuter Class: improve search capability

The chat bot can only search a specific word.

Users may not remember the specific word and to enhance user experience.

Improving the search by allowing the user to find partial words and
ignoring upper case and lower case.

Let's
* update the search engine to find partial words and ignore upper case
and lower case

guohuang88 committed:
Merge pull request #3 from guohuang88/branch-A-BetterSearch

FindExecuter Class: improve search capability

guohuang88 committed:
Update JavaDocs to adhere to proper convention

guohuang88 committed:
Update Parser class to refactor long method using SLAP

guohuang88 committed:
Update Ui class to refactor a long method using SLAP

guohuang88 committed:
Merge pull request #5 from guohuang88/branch-A-CodeQuality-2

Resolve #4 on code quality

guohuang88 committed:
Update Ui class to handle more exceptions

guohuang88 committed:
Merge pull request #6 from guohuang88/branch-A-MoreErrorHandling

Update Ui class to handle more exceptions

guohuang88 committed:
Update Main class to add a title for the GUI

guohuang88 committed:
Update README file to add user guide

guohuang88 committed:
Update FindExecutor class to improve the search

guohuang88 committed:
Update Junit test files to test methods in Parser and Ui class

guohuang88 committed:
Fix error handling issue in Ui and Parser class

guohuang88 committed:
Improve code quality for all classes in executor package

guohuang88 committed:
Update JavaDoc in all classes

guohuang88 committed:
Update the method name in UiTest class

guohuang88 committed:
Improve code quality by having better variable names

guohuang88 committed:
Update minor changes in GhBot class
